Extract dagservice and friends from go-ipfs

Stebalien opened the pull request:
This is a WIP/RFC attempt at extracting DAGService from go-ipfs. Please see the TODOs in the source files as there's a *lot* to discuss here.

However, the first question is: Is this even the right place to put this? I could create an go-ipfs-dag repo.

Stebalien commented:
Context: It turns out I will need this to get PubSub+IPLD working. IPLD without the DAG is kind of useless.

@whyrusleeping it appears I can't assign reviewers on this repo so I'm going to informally throw this on your plate :weight_lifting_man:.

@kevina You also have a fair amount of experience with this subsystem.

Kubuxu commented on merkledag.go:
```diff
@@ -0,0 +1,88 @@
+//
+// Why *not* do this? We might decide to store a light-weight DAG of links
+// without actually storing the data. I don't really find that to be a
+// convincing argument.
```
Plan with this interface is to have link cache at some time in the future, fetching object from disk to parse them to just see the links adds not that minor overhead.

Stebalien replied:
Ah. I assume this interface needs to be updated for IPLD? Non-IPFS IPLD nodes don't fill out the `Name` and `Size` fields of the `Link` structure (as a matter of fact, they *can't* accurately fill out the `Size` field).

I assume such an interface would want to return an array of:

```go
Link {
    Path []string
    Cid  *cid.Cid
}
```

(for pathing through "light-weigh" objects)

whyrusleeping replied:
The linkservice interface is useful for optimizing graph traversals. Large ipld graphs might cache links for a given node to make resolving down a path much much faster.

I agree that there might be a cleaner way, but your proposal above includes getting a node, and then calling getlinks on it. the problem there is that you have to spend time and memory unmarshaling the bytes from disk into the in memory object and then calling another function on it that returns the link array.

Stebalien replied:
>  Large ipld graphs might cache links for a given node to make resolving down a path much much faster.

Currently (I hope), this is just used for pinning/GC. The only Node type that fills in the Name field of Link structs is PBNode so this *can't* be used for resolving paths. (but it's obviously still useful or GC/pinning).

An alternative is to use upcasting/specialization (or whatever go calls it) to simplify DAGService implementations:

```go
// NodeGetters can optionally implement this interface to make 
// finding linked objects faster.
type GetLinks interface {
	NodeGetter
	GetLinks(ctx context.Context, nd *cid.Cid) ([]Link, error)
}

func GetLinks(ctx context.Context, ng NodeGetter, nd *cid.Cid) ([]Link, error) {
	if c.Type() == cid.Raw {
		return nil, nil
	}
	if gl, ok := ng.(GetLinks); ok {
		return gl.GetLinks(ctx, nd)
	}
	node, err := ng.Get(ctx, c)
	if err != nil {
		return nil, err
	}
	return node.Links(), nil
}
```

However, I don't know how prevalent this pattern is in the go-ipfs codebase and whether or not it's worth introducing.

whyrusleeping replied:
Yeah, using the upcasting/specialization pattern here sounds good to me. I also believe youre right about this only being used for GC and pinning

whyrusleeping commented on promise.go:
```diff
@@ -0,0 +1,88 @@
+// 1. NodeGetter is a naming conflict.
+// 2. It's a promise...
+
+// TODO: Should this even be an interface? It seems like a simple struct would
```
This could probably be a simple struct as opposed to an interface. I'm not sure i can see a scenario where we would want to implement this differently

Stebalien replied:
Got it.

Stebalien replied:
What are the thread-safety guarantees here? It looks like single-writer/single-reader. Is that the case?

whyrusleeping commented on merkledag.go:
```diff
@@ -0,0 +1,88 @@
+	Remove(Node) error
+
+	// TODO: This is returning them in-order?? Why not just use []NodePromise?
+	// Maybe add a couple of helpers for getting them in-order and as-available?
```
This is returned as a channel to make its usage in caller contexts much simpler. As a channel, we can just call `range` over it to do a 'for each node' type thing. Using an array of things, we would have to range over the array, then call 'get' on each inside the loop, which is slightly more complicated. Maybe we can change this in the future, but for now i think its fine

Stebalien replied:
Got it, it's just a bit inconsistent. I'll leave a shorter todo.

whyrusleeping commented on merkledag.go:
```diff
@@ -0,0 +1,88 @@
+
+// An interface for batch-adding nodes to a DAG.
+
+// TODO: Is this really the *right* level to do this at?
```
The reason for this abstraction is to avoid forcing callers to think about batching. This batch can determine the right time to flush on its own for optimal memory vs batch size tradeoffs, if done differently, each caller would have to reimplement similar logic, and that increases the chance of them doing something incorrectly

Stebalien replied:
I was thinking of defining some form of `func Batching(dag DAGService) Batch` function where `Batch` is a concrete helper type that calls `AddMany` as needed.

Pro: You don't need to implement a custom batching strategy per DAGService.
Con: You *can't* implement a custom batching strategy for individual DAGServices.

whyrusleeping replied:
I think that sounds good. Removing `Batch` as a method on the dagservice seems fine to me.

Thoughts @Kubuxu ?

Stebalien commented:
I've addressed the comments and done a little cleanup.

**To discuss**

* [ ] NodePromise thread safety guarantees.
* [ ] LinkGetter versus LinkService.

kevina commented:
Is there a p.r in go-ipfs for using this new interface?

There are several things that concern me in particular `GetOfflineLinkService` is there for a reason (of which I forget) but if things can be made to work without it than I don't object.

I would not approve of this p.r. until there is a corresponding p.r. in go-ipfs that passes all the tests.

I don't have a strong opinion either way for LinkGetter vs LinkService as long as the new interface works with go-ipfs.

whyrusleeping commented:
@kevina the suggestion for replacement is here: https://github.com/ipfs/go-ipfs/issues/4009

Stebalien commented:
@kevina, @whyrusleeping 
>  the suggestion for replacement is here: ipfs/go-ipfs#4009

Actually, for now, I've added an [`OfflineNodeGetter`](https://github.com/Stebalien/go-ipld-format/blob/54e5fc84d9fcc56cdbfb12ae8a6851545f7825bf/merkledag.go#L23) function on the `NodeGetter` interface that should serve the same purpose. Instead of passing a `LinkService` to such systems, one would pass a `NodeGetter` and call `ng.OfflineNodeGetter()` as needed. I'd like to remove this function in favor of the context based system described in that issue in the future but that will take a some design and quite a bit of refactoring.

> Is there a p.r in go-ipfs for using this new interface?

I'd rather not spend the time refactoring IPFS until we've agreed on these interfaces.

> I would not approve of this p.r. until there is a corresponding p.r. in go-ipfs that passes all the tests.

This PR (intentionally) doesn't actually touch any code that IPFS currently uses. The idea is to extract some IPLD interfaces/code from the go-ipfs repo into this repo and then later update the go-ipfs repo to depend on them (removing the duplicate code from the go-ipfs repo in the process).

kevina commented:
> I'd rather not spend the time refactoring IPFS until we've agreed on these interfaces.

I will have a hard time reviewing the interface without seeing it in use.  There are minor details that could be missed.

I could probably do the actual refactoring in a few hours.  The time consuming part for me would be the gx-publish part.

@whyrusleeping thoughts

Stebalien commented:
I'll take that as "this looks fine but my review can't progress without seeing this in use". I'll go ahead and implement that PR, I just didn't want to spend time on that without getting any feedback here (as it is, that *would* have been a waste as the code here has changed over the course of this PR).

Stebalien commented:
Actually, I should probably wait for https://github.com/ipfs/go-ipfs/pull/3990.

Kubuxu commented on merkledag.go:
```diff
@@ -0,0 +1,64 @@
+type LinkGetter interface {
+	NodeGetter
+	// TODO(ipfs/go-ipld-format#9): This should return []*cid.Cid
+	GetLinks(ctx context.Context, nd *cid.Cid) ([]*Link, error)
```
One also important reason this interface exists is that raw nodes can't have links so we don't have to fetch them to enumerate their links. This optimization was placed in LinkService.

Kubuxu replied:
but I see you placed that short-circuit in helper bellow.

Stebalien replied:
Yes; the idea is to avoid requiring that *all* DAGServices implement this method (even it they don't cache links).

Stebalien commented:
Plan:

* [ ] Get a review on the last few commits.
* [ ] Get a final sign off.
* [ ] Merge it!
* [ ] Fix all the other IPLD Node API issues all at once (annother PR)
    * [ ] #9 Remove the Link structure, it's DagProtobuf specific
    * [ ] #12 Remove the Size method
    * [ ] #10 Remove the Copy method
    * [ ] #11 Remove the Stat method
* [ ] Update all other IPLD implementations to use the now fixed IPLD Node API.
* [ ] Update go-ipfs all at once.

I originally planned on doing this in stages but I'm now of the opinion that breaking everything all at once will be much less painful (it will still be very painful...).

kevina commented:
@Stebalien the build is now failing, it looks like you forgot to commit something.

Stebalien commented:
Sorry, I'm concurrently working on the next PR (refactoring the Node API).

Stebalien commented:
I have a second PR that refactors the node API but that's blocked on finalizing the IPLD spec.

whyrusleeping commented:
link to PR? and why is it blocked on the spec? what are the open questions? (and where can we answer them)

Stebalien commented:
I haven't push it yet. I was waiting on answering the questions: Does IPLD have primitive types? If so, what? (https://github.com/ipfs/go-ipld-format/issues/25).

(This is that thing I keep on trying to schedule a discussion about).

whyrusleeping commented:
My answer was no, it doesn't have canonical types

On Wed, Oct 4, 2017, 4:59 PM Steven Allen <notifications@github.com> wrote:

> I haven't push it yet. I was waiting on answering the questions: Does IPLD
> have primitive types? If so, what? (#25
> <https://github.com/ipfs/go-ipld-format/issues/25>).
>
> (This is that thing I keep on trying to schedule a discussion about).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ipfs/go-ipld-format/pull/8#issuecomment-334164279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABL4HJKNMz2EB56wtECAmq9a5njVtMOOks5so48ngaJpZM4ODF3C>
> .
>

Stebalien commented:
David agrees with you but Juan, Nicola, the current spec (although it doesn't specify what the primitive types are), and I disagree. That's why I want to talk about it.

whyrusleeping commented:
@Stebalien can we move this forward? I need the `GetMany` interfaces for moving forward with bitswap session integrations.

Stebalien commented:
@whyrusleeping You can always add that to the copy of `DAGService` in go-ipfs if you want to get that integration done faster.

However, we should probably just move forward with this anyways (and leave fixing the node API to a future PR).

I have a few more questions open for discussion but we can punt on those if you want.

Stebalien commented on batch.go:
```diff
@@ -0,0 +1,113 @@
+package format
```
This is a *lot* of code for an interface package. Unfortunately, the interface needs this (unless we decide to introduce a `Batch` interface.

Stebalien commented on merkledag.go:
```diff
@@ -0,0 +1,58 @@
+	NodeGetter
+
+	// Add adds a node to this DAG.
+	Add(Node) (*cid.Cid, error)
```
Does this really need to return the Cid? The caller can also just call `node.Cid()`.

whyrusleeping replied:
True

Kubuxu replied:
I agree

Stebalien commented on merkledag.go:
```diff
@@ -0,0 +1,58 @@
+	//
+	// Consider using NewBatch instead of calling this directly if you need
+	// to add an unbounded number of nodes to avoid buffering too much.
+	AddMany([]Node) ([]*cid.Cid, error)
```
Same here. Worse, this will allocate something that we're likely to ignore (that's why I bothered bringing it up).

whyrusleeping replied:
Probably fine to not have it return cids, as you say, though there might be cases where we need an array of cids directly after calling AddMany, and that will kindof suck.

Stebalien replied:
IMO, we can just create a `Cids([]Node) []*cid.Cid` function for that.

whyrusleeping replied:
Yeah, that SGTM

Kubuxu replied:
I agree also.

Stebalien commented on merkledag.go:
```diff
@@ -0,0 +1,58 @@
+	// Remove removes a node from this DAG.
+	//
+	// If the node is not in this DAG, Remove returns ErrNotFound.
+	Remove(*cid.Cid) error
```
Do we also want `RemoveMany(...)` (for, e.g., dag-level GC)?

whyrusleeping replied:
We can probably punt on this

Kubuxu replied:
IMO yes.

Stebalien commented on merkledag.go:
```diff
@@ -0,0 +1,58 @@
+}
+
+// DAGService is an IPFS Merkle DAG service.
+type DAGService interface {
```
Should all of these methods take contexts? Currently, they don't because they never hit the network. However, I assume we'll eventually want remote DAGs.

Kubuxu replied:
We might want to add it now.

Stebalien commented:
Actually, I'm just going to punt on all of those questions and merge this if I don't get any objections within the next few hours. 🚀

Kubuxu commented:
I wanted to give it a review tomorrow morning.

Stebalien commented:
@Kubuxu got it. I'll hold off. FYI, I've decided to punt on fixing the Node API till later so we can get stop blocking this on "one more question".

Kubuxu commented on format.go:
```diff
@@ -19,6 +19,10 @@ type Resolver interface {
+// Node is the base interface all IPLD nodes must implement.
+//
+// Nodes are **Immutable** and all methods defined on the interface are
+// **Thread Safe**.
```
If Nodes are immutable, what is the reason for `Copy` method?

Stebalien replied:
None. I was planning on fixing that in this PR but have punted on fixing Nodes till a future one.

Stebalien commented:
@Kubuxu I've decided to just go ahead and address all your comments/the todos *now*.

* Added contexts.
* No longer returning CIDs from Add.
* Added RemoveMany.

Also, improved Batch by adding context support (we'll need that to use batch from the API).

However, this is starting to become an ever growing PR so I should probably stop now... Final review?

whyrusleeping commented:
If we're on the breaking interfaces train already, i'd love to get a `Load(ctx, cid, &thing)` method on the dagservice... Its better for memory usage and it makes dealing with cbor things much easier.

Also fine having that be in a follow-up too though

Stebalien commented:
@whyrusleeping we could also just lazily deserialize nodes. Are you worried about the intermediate buffer?

Stebalien commented:
@Kubuxu IIRC, you wanted to do a final review.

kevina commented:
This p.r. sorta interferes with my effort to create better error messages, which I believe will require some API changes to make it happen.  I would like to work though https://github.com/ipfs/go-ipfs/pull/4468 first as it is easier to modify things when they are all in the same package.  But if this is blocking something I don't want to hold anything in,espacally considering I just created the p.r. yesterday.

Stebalien commented:
@kevina I'd like to push this through ASAP. @whyrusleeping wants some of these changes and this has been sitting here for months.
